classes/mail/Mailer.inc.php

```diff
@@ -0,0 +1,196 @@
+     */
+    protected function renderView($view, $data) : string
+    {
+        if ($view instanceof Htmlable)
+            // return HTML without data compiling
+            return $view->toHtml();
+
+        if (!is_string($view))
+            throw new InvalidArgumentException('View must be instance of ' . Htmlable::class . ' or a string, ' . get_class($view) . ' is given');
+
+        return $this->compileParams($view, $data);
+    }
+
+    /**
+     * @param string $view text or HTML string
+     * @param array $data variables with their values passes ['variable' => value]
+     * @return string compiled string with substitute variables
+     * @throws Exception
+     */
+    public function compileParams(string $view, array $data, string $locale = null) : string
+    {
+        $data = $this->localizeData($data, $locale);
+        unset($data[Mailable::DATA_KEY_MESSAGE]); // Remove pre-set message variable
+
+        $stack = [];
+        $resultString = '';
+        $variable = '';
+        foreach ($template = str_split($view) as $key => $char) {
+            if ($this->openTags($key, $char, $template)) {
+                if (empty($stack)) {
+                    $stack[] = $char;
+                } else {
+                    throw new Exception('Syntax error in email template markdown: opening a new variable tag without closing previous one');
+                }
+                continue;
+            }
+
+            if ($char === self::CLOSING_TAG && !empty($stack)) {
+                array_pop($stack);
+                $resultString = $this->recordVariable($resultString, $variable, $data);
+                $variable = '';
+                continue;
+            }
+
+            if (!empty($stack)) {
+                $variable .= $char;
+                continue;
+            }
+
+            $resultString .= $char;
+        }
+
+        if (!empty($stack))
+            throw new Exception('Template variable tag should be closed');
+
+        return $resultString;
+    }
+
+    /**
+     * @param array $data
+     * @param string|null $locale
+     * @return array localized data
+     */
+    protected function localizeData(array $data, string $locale = null) : array
+    {
+        if (is_null($locale))
+            $locale = PKPLocale::getLocale();
+
+        if (!PKPLocale::isLocaleValid($locale))
+            throw new InvalidArgumentException($locale . ' isn\'t recognized as a valid locale');
+
+        return array_map(function ($varValue) use ($locale) {
+
+            // Array values contain localized template data
+            if (is_array($varValue)) {
+                if (array_key_exists($locale, $varValue)) {
+                    return $varValue[$locale];
+                } else {
+                    throw new InvalidArgumentException('Template variables doesn\'t have localization for the ' . $locale . ' locale' );
+                }
+            }
+            return $varValue;
+        }, $data);
+    }
```

Comment:
Hmm, let's talk about the use cases behind this. I think we're going to run into trouble trying to automate the localization process outside of the objects -- localization has specific logic around fallbacks when locale data is missing and I think we need to preserve those.

My initial thought is that we probably need to define a locale when the mailable is instantiated, or before the email variable classes are used, and the use each object's `getLocalizedData()` methods to return the correct value (with appropriate fallbacks).

Reply:
This shouldn't prevent a merge, though.

Reply:
https://github.com/pkp/pkp-lib/issues/7258